File: Testio/app/src/main/java/com/azzdorfrobotics/android/testio/network/NetworkModule.java
package com.azzdorfrobotics.android.testio.network;

import android.content.Context;
import android.net.ConnectivityManager;
import com.azzdorfrobotics.android.testio.network.model.NetworkState;
import com.azzdorfrobotics.android.testio.preferences.Key;
import com.google.gson.Gson;
import dagger.Module;
import dagger.Provides;
import java.util.concurrent.TimeUnit;
import javax.inject.Named;
import javax.inject.Singleton;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created on 25.12.2017.
 *
 * @author Mykola Tychyna (iMykolaPro)
 */

@Module public class NetworkModule {

    @Provides @Singleton Gson provideGson() {
        /**
         * Custom converters have to be added here
         */
        return new Gson();
    }

    @Provides @Singleton GsonConverterFactory provideGsonConverterFactory(Gson gson) {
        return GsonConverterFactory.create(gson);
    }

    @Provides @Singleton OkHttpClient provideHttpClient(
        TesonetHeaderInterceptor tesonetHeaderInterceptor) {
        OkHttpClient.Builder builder = new OkHttpClient.Builder();

        builder.addInterceptor(tesonetHeaderInterceptor);

        HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
        logging.setLevel(HttpLoggingInterceptor.Level.BODY);
        builder.addInterceptor(logging);

        builder.readTimeout(20, TimeUnit.SECONDS);
        builder.writeTimeout(20, TimeUnit.SECONDS);
        builder.connectTimeout(20, TimeUnit.SECONDS);

        return builder.build();
    }

    @Provides @Singleton TesonetApi provideTesonetApi(@Named(Key.API_URL) String url,
        OkHttpClient okHttpClient, GsonConverterFactory gsonConverterFactory) {
        return new Retrofit.Builder().baseUrl(url)
            .client(okHttpClient)
            .addConverterFactory(gsonConverterFactory)
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .build()
            .create(TesonetApi.class);
    }

    @Provides @Singleton ConnectivityManager provideConnectivityManager(Context context) {
        return (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
    }

    @Provides @Singleton NetworkState provideNetworkState(
        ConnectivityManager connectivityManagerCompat) {
        return new NetworkState(connectivityManagerCompat);
    }
}

File: app/src/main/java/com/teso/net/di/NetworkModule.kt
package com.teso.net.di

import com.facebook.stetho.okhttp3.StethoInterceptor
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.teso.net.BuildConfig
import com.teso.net.data_flow.interactions.ITokenInteractor
import com.teso.net.data_flow.network.Api
import com.teso.net.data_flow.network.TokenInterceptor
import dagger.Module
import dagger.Provides
import okhttp3.HttpUrl
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import timber.log.Timber
import java.util.concurrent.TimeUnit
import javax.inject.Singleton

@Module
class NetworkModule {

    @Singleton
    @Provides
    fun provideApiInterface(retrofit: Retrofit): Api = retrofit.create(Api::class.java)


    @Singleton
    @Provides
    fun provideBaseRetrofit(httpUrl: HttpUrl,
                            client: OkHttpClient,
                            gson: Gson): Retrofit {
        return Retrofit.Builder()
                .baseUrl(httpUrl)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.createAsync())
                .build()
    }

    @Singleton
    @Provides
    fun provideHttpUrl() = HttpUrl.parse(BuildConfig.API_BASE_URL)!!

    @Singleton
    @Provides
    fun provideOkHttpClient(loggerInterceptor: HttpLoggingInterceptor,
                            tokenInterceptor: TokenInterceptor,
                            headersInterceptor: Interceptor): OkHttpClient = OkHttpClient.Builder()
            .addInterceptor(headersInterceptor)
            .addInterceptor(tokenInterceptor)
            .addNetworkInterceptor(loggerInterceptor)
            .addNetworkInterceptor(StethoInterceptor())
            .connectTimeout(20, TimeUnit.SECONDS)
            .readTimeout(20, TimeUnit.SECONDS)
            .writeTimeout(20, TimeUnit.SECONDS)
            .build()

    @Singleton
    @Provides
    fun provideHttpLoggerInterceptor(): HttpLoggingInterceptor {
        val logger = HttpLoggingInterceptor(HttpLoggingInterceptor.Logger {
            Timber.tag(NETWORK_TAG).d(it)
        })
        logger.level = if (BuildConfig.DEBUG) HttpLoggingInterceptor.Level.BODY else HttpLoggingInterceptor.Level.NONE

        return logger
    }

    @Singleton
    @Provides
    fun provideTokenInterceptor(loginInteractor: ITokenInteractor): TokenInterceptor {
        return TokenInterceptor(loginInteractor)
    }

    @Singleton
    @Provides
    fun provideHeadersInterceptor() = Interceptor {
        it.proceed(
                it.request().newBuilder()
                        .addHeader("Accept", "application/json")
                        .addHeader("Accept-Language", "en")
                        .addHeader("Content-Type", "application/json")
                        .build()
        )
    }

    @Singleton
    @Provides
    fun provideGson(): Gson = GsonBuilder()
            .serializeNulls()
            .create()

    companion object {
        private const val NETWORK_TAG = "Network"
    }
}